src/backend/distributed/ddl/foreign_constraint.c

```diff
@@ -21,11 +22,588 @@
+	{
+		frelGraph->transitivityMatrix[curIndex] = (bool *) palloc(frelGraph->nodeCount *
+																  sizeof(bool));
+		memset(frelGraph->transitivityMatrix[curIndex], false, frelGraph->nodeCount *
+			   sizeof(bool));
+		curIndex += 1;
+	}
+
+	hash_seq_init(&status, frelGraph->nodeMap);
+
+	/*
+	 * Create the initial adjacency matrix, that will be transformed to the
+	 * transitivity matrix dynamically.
+	 */
+	while ((frelNode = (FRelNode *) hash_seq_search(&status)) != 0)
+	{
+		uint32 relationIndex = frelNode->index;
+		ListCell *nodeCell = NULL;
+		List *curAdjacencyList = frelNode->adjacencyList;
+
+
+		foreach(nodeCell, curAdjacencyList)
+		{
+			FRelNode *referencingFrelNode = (FRelNode *) lfirst(nodeCell);
+			uint32 referencingRelationIndex = referencingFrelNode->index;
+			frelGraph->transitivityMatrix[relationIndex][referencingRelationIndex] = true;
+		}
+	}
+
+	/*
+	 * Add all vertices one by one to the set of intermediate vertices.
+	 *
+	 * Before start of a iteration, we have transitivity values for all
+	 * pairs of vertices such that the transitivity values consider only
+	 * the vertices in set {0, 1, 2, .. k-1} as intermediate vertices.
+	 *
+	 * After the end of a iteration, vertex no. k is added to the set of
+	 * intermediate vertices and the set becomes {0, 1, .. k}
+	 */
+	for (interVertexCounter = 0; interVertexCounter < frelGraph->nodeCount;
+		 interVertexCounter++)
+	{
+		/* pick all vertices as source one by one */
+		for (sourceCounter = 0; sourceCounter < frelGraph->nodeCount; sourceCounter++)
+		{
+			/* pick all vertices as destination for the above picked source */
+			for (destinationCounter = 0; destinationCounter < frelGraph->nodeCount;
+				 destinationCounter++)
+			{
+				/*
+				 * If vertex k is on a path from i to j, then make sure that
+				 * the value of transitivityMatrix[i][j] is 1.
+				 */
+				frelGraph->transitivityMatrix[sourceCounter][destinationCounter] =
+					frelGraph->transitivityMatrix[sourceCounter][destinationCounter] ||
+					(frelGraph->transitivityMatrix[sourceCounter][interVertexCounter] &&
+					 frelGraph->transitivityMatrix[interVertexCounter][destinationCounter]);
+			}
+		}
+	}
+}
+
+
+/*
+ * ClearForeignKeyRelationGraph clear all the allocated memory obtained for
+ * foreign key relation graph.
+ */
+void
+ClearForeignKeyRelationGraph()
+{
+	MemoryContext oldContext = MemoryContextSwitchTo(CacheMemoryContext);
+	int indexCounter = 0;
+	if (frelGraph == NULL)
+	{
+		return;
+	}
+
+	/*
+	 * Free the transitivity matrix and index-oid mapping array of foreign key
+	 * relation graph. If graph doesn't have any node, then we do not have neither
+	 * of them.
+	 */
+	if (frelGraph->nodeCount > 0)
+	{
+		for (indexCounter = 0; indexCounter < frelGraph->nodeCount; indexCounter++)
+		{
+			pfree(frelGraph->transitivityMatrix[indexCounter]);
+		}
+
+		pfree(frelGraph->indexToOidArray);
+	}
+
+	/* free the map holding relation nodes */
+	hash_destroy(frelGraph->nodeMap);
```

Comment:
I'm pretty sure this will clear the entire `CacheMemoryContext`, since `hash_destroy` clears the hash's memory context and you've set it to `CurrentMemoryContext` rather than a child.